### PR TITLE
ENH: allow .rolling / .expanding as groupby methods

### LINKED ISSUE #12363: BUG: inconsistent name for returned Series in groupby
I would think ``[7]`` and ``[8]`` would be identical. Note this is true for all ops I have looked at, so it is a general issue.

```
In [6]:         df = DataFrame({'A': ['foo', 'bar', 'foo', 'bar',
                              'foo', 'bar', 'foo', 'foo'],
                        'B': ['one', 'one', 'two', 'two',
                              'two', 'two', 'one', 'two'],
                        'C': np.random.randn(8) + 1.0,
                        'D': np.arange(8)})

In [7]: df.groupby(['A']).B.count()
Out[7]: 
A
bar    3
foo    5
Name: B, dtype: int64

In [8]: df.B.groupby(df.A).count()
Out[8]: 
A
bar    3
foo    5
dtype: int64
```

### LINKED ISSUE #12486: API: groupby.resample *maybe* can return a deferred operation
xref #12448 / #12449 

and on [SO](http://stackoverflow.com/questions/36293620/pandas-0-18-changes-to-resample-how-to-upsample-with-groupby/36294398#36294398)

```
In [1]:         df = DataFrame({'date': pd.date_range(start='2016-01-01',
   ...:                                               periods=4,
   ...:                                               freq='W'),
   ...:                         'group': [1, 1, 2, 2],
   ...:                         'val': [5, 6, 7, 8]}).set_index('date')

In [2]: df
Out[2]: 
            group  val
date                  
2016-01-03      1    5
2016-01-10      1    6
2016-01-17      2    7
2016-01-24      2    8
```

This replicates 0.17.1 (something slightly off with it including the grouper column)
```
In [3]: df.groupby('group').apply(lambda x: x.resample('1D').ffill())[['val']]
Out[3]: 
                  val
group date           
1     2016-01-03    5
      2016-01-04    5
      2016-01-05    5
      2016-01-06    5
      2016-01-07    5
      2016-01-08    5
      2016-01-09    5
      2016-01-10    6
2     2016-01-17    7
      2016-01-18    7
      2016-01-19    7
      2016-01-20    7
      2016-01-21    7
      2016-01-22    7
      2016-01-23    7
      2016-01-24    8

# ideally this would work. Its possible but requires some intelligently filling according to each group level.
In [4]: df.groupby('group').resample('1D').ffill()
Out[4]: 
            group  val
date                  
2016-01-03      1    5
2016-01-10      1    6
2016-01-17      2    7
```

A pure asfreq operation
```
data = [['2010-01-01', 'A', 2], ['2010-01-02', 'A', 3], ['2010-01-05', 'A', 8], 
        ['2010-01-10', 'A', 7], ['2010-01-13', 'A', 3], ['2010-01-01', 'B', 5], 
        ['2010-01-03', 'B', 2], ['2010-01-04', 'B', 1], ['2010-01-11', 'B', 7], 
        ['2010-01-14', 'B', 3]]

df = pd.DataFrame(data, columns=['Date', 'ID', 'Score'])
df.Date = pd.to_datetime(df.Date)

In [27]: df.groupby('ID').apply(lambda x: x.set_index('Date').Score.resample('D').asfreq())
Out[27]: 
ID  Date      
A   2010-01-01    2.0
    2010-01-02    3.0
    2010-01-03    NaN
    2010-01-04    NaN
    2010-01-05    8.0
    2010-01-06    NaN
    2010-01-07    NaN
    2010-01-08    NaN
    2010-01-09    NaN
    2010-01-10    7.0
    2010-01-11    NaN
    2010-01-12    NaN
    2010-01-13    3.0
B   2010-01-01    5.0
    2010-01-02    NaN
    2010-01-03    2.0
    2010-01-04    1.0
    2010-01-05    NaN
    2010-01-06    NaN
    2010-01-07    NaN
    2010-01-08    NaN
    2010-01-09    NaN
    2010-01-10    NaN
    2010-01-11    7.0
    2010-01-12    NaN
    2010-01-13    NaN
    2010-01-14    3.0
Name: Score, dtype: float64
```

Would be nice for this to work
```
df.groupby(['ID',pd.Grouper(key='Date',freq='D')]).asfreq()
```

### LINKED ISSUE #12738: Allow .expanding() to be called on groupby object
Right now when I want to have expanding operate at the level of a group I have to apply a convenience function like so:

```python
def expanding_sum(s):
    return s.expanding().sum()

expanded_sum = df.groupby('mygroup')['mynum'].apply(expanding_sum)
```

Would be nice to be able to call a method directly on a groupby object:

```python
expanded_sum = df.groupby('mygroup')['mynum'].expanding().sum()
```

### PR DESCRIPTION
**Pull request by jreback**
closes #12738
closes #12486 
closes #12363
- [x] more tests (other methods)

~~\- [ ] doc section in groupby~~ will do later

```
In [3]: pd.options.display.max_rows=10

In [4]:    df = pd.DataFrame({'A': [1] * 20 + [2] * 12 + [3] * 8,
                      'B': np.arange(40)})

In [5]: df
Out[5]: 
    A   B
0   1   0
1   1   1
2   1   2
3   1   3
4   1   4
.. ..  ..
35  3  35
36  3  36
37  3  37
38  3  38
39  3  39

[40 rows x 2 columns]

In [6]:    df.groupby('A').apply(lambda x: x.rolling(4).B.mean())
Out[6]: 
A    
1  0      NaN
   1      NaN
   2      NaN
   3      1.5
   4      2.5
         ... 
3  35    33.5
   36    34.5
   37    35.5
   38    36.5
   39    37.5
Name: B, dtype: float64

In [7]:    df.groupby('A').rolling(4).B.mean()
Out[7]: 
A    
1  0      NaN
   1      NaN
   2      NaN
   3      1.5
   4      2.5
         ... 
3  35    33.5
   36    34.5
   37    35.5
   38    36.5
   39    37.5
Name: B, dtype: float64
In [9]:    df.index = pd.date_range('20130101',freq='s',periods=40)

In [10]:    df
Out[10]: 
                     A   B
2013-01-01 00:00:00  1   0
2013-01-01 00:00:01  1   1
2013-01-01 00:00:02  1   2
2013-01-01 00:00:03  1   3
2013-01-01 00:00:04  1   4
...                 ..  ..
2013-01-01 00:00:35  3  35
2013-01-01 00:00:36  3  36
2013-01-01 00:00:37  3  37
2013-01-01 00:00:38  3  38
2013-01-01 00:00:39  3  39

[40 rows x 2 columns]

In [11]:    df.groupby('A').apply(lambda x: x.resample('4s').mean())
Out[11]: 
                         A     B
A                               
1 2013-01-01 00:00:00  1.0   1.5
  2013-01-01 00:00:04  1.0   5.5
  2013-01-01 00:00:08  1.0   9.5
  2013-01-01 00:00:12  1.0  13.5
  2013-01-01 00:00:16  1.0  17.5
2 2013-01-01 00:00:20  2.0  21.5
  2013-01-01 00:00:24  2.0  25.5
  2013-01-01 00:00:28  2.0  29.5
3 2013-01-01 00:00:32  3.0  33.5
  2013-01-01 00:00:36  3.0  37.5

In [12]:    df.groupby('A').resample('4s').mean()
Out[12]: 
                         A     B
A                               
1 2013-01-01 00:00:00  1.0   1.5
  2013-01-01 00:00:04  1.0   5.5
  2013-01-01 00:00:08  1.0   9.5
  2013-01-01 00:00:12  1.0  13.5
  2013-01-01 00:00:16  1.0  17.5
2 2013-01-01 00:00:20  2.0  21.5
  2013-01-01 00:00:24  2.0  25.5
  2013-01-01 00:00:28  2.0  29.5
3 2013-01-01 00:00:32  3.0  33.5
  2013-01-01 00:00:36  3.0  37.5
```

### CODE REVIEW

**Comment by jreback** [comment]
cc @lminer

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -794,6 +794,7 @@ def _concat_objects(self, keys, values, not_indexed_same=False):
 
             if isinstance(result, Series):
                 result = result.reindex(ax)
+                result.name = self.name
```
Comment (jreback) [inline pandas/core/groupby.py]: @sinhrks this fixed a couple of tests in groupby below, but I don't know if we have a related issue.

related (but did not fix) #12363

File: pandas/tests/test_groupby.py [hunk 2]
```diff
@@ -4119,6 +4117,7 @@ def test_groupby_multi_timezone(self):
                                      tz='America/Chicago'),
                            Timestamp('2000-01-01 16:50:00-0500',
                                      tz='America/New_York')],
+                          name='date',
```
Comment (jreback) [inline pandas/tests/test_groupby.py]: e.g. this was fixed

File: pandas/core/groupby.py (hunk 1 above)
Comment (sinhrks) [inline pandas/core/groupby.py]: Not sure, I can find no open issue.

**Comment by jreback** [comment]
ok, this is ready. comments.

@shoyer @jorisvandenbossche @TomAugspurger @sinhrks

**Comment by jorisvandenbossche** [comment]
Is the example at the top still up to date? 
I am wondering a bit if the grouper should be in the result or not?

**Comment by jorisvandenbossche** [comment]
Also, how does the result looks like if you have a non-sorted grouper? 
Currently with apply it sorts the groups (but also includes the grouper as the index)

**Comment by jreback** [comment]
@jorisvandenbossche all updated. Had to work thru some issues. But much more fully tested now. This basically replicates what ``.apply`` is doing with the new syntax by actually *using* apply. In theory we could have a more efficient impl, but will leave that for later.

Also this PR cleans up a bunch of cases where the name is not returned for groupbys. So much more consistency now.

File: pandas/core/groupby.py [hunk 3]
```diff
@@ -339,16 +342,23 @@ def __init__(self, obj, keys=None, axis=0, level=None,
         self.sort = sort
         self.group_keys = group_keys
         self.squeeze = squeeze
+        self.mutated = kwargs.pop('mutated', False)
```
Comment (jreback) [inline pandas/core/groupby.py]: this is a massive hack using. well hack is the wrong word. Its a way of informing the groupby that we want to force the multiindex construction path which is normally taken only when things are mutated. its not an external visible kw.

File: pandas/tests/test_window.py [hunk 4]
```diff
@@ -2713,3 +2713,129 @@ def test_rolling_min_max_numeric_types(self):
...
+        g = self.frame.groupby('A')
+        self.assertFalse(g.mutated)
+        g = self.frame.groupby('A', mutated=True)
+        self.assertTrue(g.mutated)
+
+    def test_getitem(self):
+        g = self.frame.groupby('A')
+        g_mutated = self.frame.groupby('A', mutated=True)
+
+        expected = g_mutated.B.apply(lambda x: x.rolling(2).mean())
```
Comment (jreback) [inline pandas/tests/test_window.py]: so by our current rules this *wont'* return a multi-index (as its a regular groupby with a series returned, which effectively just get stacked). So for testing I *have* to return a multi-index which is the proper result. Note this is a ``Series`` groupby here (if you do the ``DataFrame`` it returns a multi-index.

This ``.apply`` DID not work in 0.18.0 (for a DataFrame), e.g.

This raised
``df.groupby('A').apply(lambda x: x.rolling(2).sum())``

This worked
``df.groupby('A').B.apply(lambda x: x.rolling(2).sum())``

So I kept it the same. 

cc @jorisvandenbossche

Comment (jreback) [inline pandas/tests/test_window.py]: The fundamental thing is its very tricky to infer whether to create mutli-indexes with a user passed function. So I think that we 'guess' by seeing if it is indexed the same or not (e.g. mutated). Can't really fault the logic. With the new ``.rolling/expanding/resample`` we KNOW that we need to produce a MI.

**Comment by jreback** [comment]
@jorisvandenbossche if you'd have a look

**Comment by jorisvandenbossche** [comment]
Will take a look later today!

File: doc/source/whatsnew/v0.18.1.txt [hunk 5]
```diff
@@ -48,6 +50,48 @@ see :ref:`Custom Business Hour <timeseries.custombusinesshour>` (:issue:`11514`)
...
+
+``.groupby(...)`` has been enhanced to provide convenient syntax when working with ``.rolling(..)``, ``.expanding(..)`` and ``.resample(..)`` per group, see (:issue:`12486`, :issue:`12738`).
+
+.. ipython:: python
+
+   df = pd.DataFrame({'A': [1] * 20 + [2] * 12 + [3] * 8,
+                      'B': np.arange(40)})
+   df
+
+You can now use ``.rolling(..)`` and ``.expanding(..)`` as methods on groupbys. These return another object where you operate.
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.18.1.txt]: "These return another object where you operate" -> this is not really a clear sentence. What do exactly want to say?

File: pandas/core/window.py [hunk 6]
```diff
@@ -857,6 +884,18 @@ def corr(self, other=None, pairwise=None, **kwargs):
                                          **kwargs)
 
 
+class RollingGroupby(_GroupByMixin, Rolling):
+    """
+    Provides a rolling groupby implementation
+
+    .. versionadded:: 0.18.0
```
Comment (jorisvandenbossche) [inline pandas/core/window.py]: 0.18.1

File: pandas/core/window.py [hunk 7]
```diff
@@ -1005,6 +1044,18 @@ def corr(self, other=None, pairwise=None, **kwargs):
                                            **kwargs)
 
 
+class ExpandingGroupby(_GroupByMixin, Expanding):
+    """
+    Provides a expanding groupby implementation
+
+    .. versionadded:: 0.18.0
```
Comment (jorisvandenbossche) [inline pandas/core/window.py]: idem

File: pandas/tests/test_window.py (hunk 4 above)
Comment (jorisvandenbossche) [inline pandas/tests/test_window.py]: But for a DataFrame it still does not give a MI?

File: doc/source/whatsnew/v0.18.1.txt [hunk 8]
```diff
@@ -48,6 +50,48 @@ see :ref:`Custom Business Hour <timeseries.custombusinesshour>` (:issue:`11514`)
...
+
+   df.groupby('A').apply(lambda x: x.rolling(4).B.mean())
+
+Now you can do:
+
+.. ipython:: python
+
+   df.groupby('A').rolling(4).B.mean()
+
+For ``.resample(..)`` type of operations, previously you would have to:
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.18.1.txt]: ` df.groupby('A').resample('4s', 'mean')` already worked previously I think? And in 0.18.0 also `df.groupby('A').resample('4s').mean()` already worked.

**Comment by jorisvandenbossche** [comment]
Still trying to wrap my head around these changes, but already notices two things (using the timeseries df from the whatsnew docs):

* Backwards compatibility for resample on groupby. With this PR this now gives a ResamplerGroupby:

  ```
In [33]: df.groupby('A').resample('4s', 'mean')
Out[33]: DatetimeIndexResamplerGroupby [freq=<4 * Seconds>, axis=0, closed=left,
 label=left, convention=e, base=0]

  ```

  While with 0.18.0 this still gives you a deprecation warning and the same result as in 0.17:

  ```
In [16]:  df.groupby('A').resample('4s', 'mean')
C:\Anaconda\envs\pandas018\Scripts\ipython-script.py:1: FutureWarning: how in .r
esample() is deprecated
the new syntax is .resample(...).mean()
  if __name__ == '__main__':
Out[16]:
                       A     B
A
1 2013-01-01 00:00:00  1   1.5
  2013-01-01 00:00:04  1   5.5
  2013-01-01 00:00:08  1   9.5
  2013-01-01 00:00:12  1  13.5
  2013-01-01 00:00:16  1  17.5
2 2013-01-01 00:00:20  2  21.5
  2013-01-01 00:00:24  2  25.5
  2013-01-01 00:00:28  2  29.5
3 2013-01-01 00:00:32  3  33.5
  2013-01-01 00:00:36  3  37.5
  ```

* There seems to be an inconsistency between returning a MI between resample and rolling (for dataframes):

  ```
In [30]: df.groupby('A').resample('3s').mean()
Out[30]:
                       A     B
A
1 2013-01-01 00:00:00  1   1.0
  2013-01-01 00:00:03  1   4.0
  2013-01-01 00:00:06  1   7.0
  2013-01-01 00:00:09  1  10.0
  2013-01-01 00:00:12  1  13.0
  2013-01-01 00:00:15  1  16.0
  2013-01-01 00:00:18  1  18.5
2 2013-01-01 00:00:18  2  20.0
  2013-01-01 00:00:21  2  22.0
  2013-01-01 00:00:24  2  25.0
  2013-01-01 00:00:27  2  28.0
  2013-01-01 00:00:30  2  30.5
3 2013-01-01 00:00:30  3  32.0
  2013-01-01 00:00:33  3  34.0
  2013-01-01 00:00:36  3  37.0
  2013-01-01 00:00:39  3  39.0

In [31]: df.groupby('A').rolling(3).mean()
Out[31]:
                       A     B
2013-01-01 00:00:00  NaN   NaN
2013-01-01 00:00:01  NaN   NaN
2013-01-01 00:00:02  1.0   1.0
2013-01-01 00:00:03  1.0   2.0
2013-01-01 00:00:04  1.0   3.0
2013-01-01 00:00:05  1.0   4.0
2013-01-01 00:00:06  1.0   5.0
2013-01-01 00:00:07  1.0   6.0
2013-01-01 00:00:08  1.0   7.0
2013-01-01 00:00:09  1.0   8.0
2013-01-01 00:00:10  1.0   9.0
2013-01-01 00:00:11  1.0  10.0
2013-01-01 00:00:12  1.0  11.0
2013-01-01 00:00:13  1.0  12.0
2013-01-01 00:00:14  1.0  13.0
...
  ```

  (in case of selecting a series (` df.groupby('A').B.rolling(3).mean()` and `df.groupby('A').B.resample('3s').mean()` both return a MI)

File: pandas/tests/test_window.py (hunk 4 above)
Comment (jreback) [inline pandas/tests/test_window.py]: sure it does
```
df = pd.DataFrame({'A': [1] * 20 + [2] * 12 + [3] * 8,
                      'B': np.arange(40)})

In [6]: df.groupby('A').apply(lambda x: x.rolling(4).mean())
Out[6]: 
        A     B
A              
1 0   NaN   NaN
  1   NaN   NaN
  2   NaN   NaN
  3   1.0   1.5
  4   1.0   2.5
...   ...   ...
3 35  3.0  33.5
  36  3.0  34.5
  37  3.0  35.5
  38  3.0  36.5
  39  3.0  37.5

[40 rows x 2 columns]

In [7]: df.groupby('A').rolling(4).mean()
Out[7]: 
        A     B
A              
1 0   NaN   NaN
  1   NaN   NaN
  2   NaN   NaN
  3   1.0   1.5
  4   1.0   2.5
...   ...   ...
3 35  3.0  33.5
  36  3.0  34.5
  37  3.0  35.5
  38  3.0  36.5
  39  3.0  37.5

[40 rows x 2 columns]
```

File: doc/source/whatsnew/v0.18.1.txt (hunk 5 above)
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: updated

File: pandas/core/window.py (hunk 6 above)
Comment (jreback) [inline pandas/core/window.py]: fixed

File: pandas/core/window.py (hunk 7 above)
Comment (jreback) [inline pandas/core/window.py]: done

File: doc/source/whatsnew/v0.18.1.txt (hunk 8 above)
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: it worked partially, but only for downsampling. I'll give a different example which illustrates.

**Comment by jreback** [comment]
on your last, can you show a complete example. this works for me

```
In [5]: df = DataFrame({'A': [1] * 20 + [2] * 12 + [3] * 8,
                                'B': np.arange(40)},
                               index=date_range('1/1/2000',
                                                freq='s',
                                                periods=40))

In [6]: df
Out[6]: 
                     A   B
2000-01-01 00:00:00  1   0
2000-01-01 00:00:01  1   1
2000-01-01 00:00:02  1   2
2000-01-01 00:00:03  1   3
2000-01-01 00:00:04  1   4
...                 ..  ..
2000-01-01 00:00:35  3  35
2000-01-01 00:00:36  3  36
2000-01-01 00:00:37  3  37
2000-01-01 00:00:38  3  38
2000-01-01 00:00:39  3  39

[40 rows x 2 columns]

In [7]: df.groupby('A').resample('3s').mean()
Out[7]: 
                         A     B
A                               
1 2000-01-01 00:00:00  1.0   1.0
  2000-01-01 00:00:03  1.0   4.0
  2000-01-01 00:00:06  1.0   7.0
  2000-01-01 00:00:09  1.0  10.0
  2000-01-01 00:00:12  1.0  13.0
...                    ...   ...
2 2000-01-01 00:00:30  2.0  30.5
3 2000-01-01 00:00:30  3.0  32.0
  2000-01-01 00:00:33  3.0  34.0
  2000-01-01 00:00:36  3.0  37.0
  2000-01-01 00:00:39  3.0  39.0

[16 rows x 2 columns]

In [8]: df.groupby('A').rolling(3).mean()
Out[8]: 
                         A     B
A                               
1 2000-01-01 00:00:00  NaN   NaN
  2000-01-01 00:00:01  NaN   NaN
  2000-01-01 00:00:02  1.0   1.0
  2000-01-01 00:00:03  1.0   2.0
  2000-01-01 00:00:04  1.0   3.0
...                    ...   ...
3 2000-01-01 00:00:35  3.0  34.0
  2000-01-01 00:00:36  3.0  35.0
  2000-01-01 00:00:37  3.0  36.0
  2000-01-01 00:00:38  3.0  37.0
  2000-01-01 00:00:39  3.0  38.0

[40 rows x 2 columns]
```

**Comment by jreback** [comment]
@jorisvandenbossche all fixed up (was not doing the deprecation step).

**Comment by jorisvandenbossche** [comment]
Just fetched your PR again, and I am still seeing this difference:

```
In [1]: df = pd.DataFrame({'A': [1] * 20 + [2] * 12 + [3] * 8,'B': np.arange(40)})

In [2]: df.index = pd.date_range('20130101',freq='s',periods=40)

In [3]: df.groupby('A').resample('3s').mean()
Out[3]:
                       A     B
A
1 2013-01-01 00:00:00  1   1.0
  2013-01-01 00:00:03  1   4.0
  2013-01-01 00:00:06  1   7.0
  2013-01-01 00:00:09  1  10.0
  2013-01-01 00:00:12  1  13.0
  2013-01-01 00:00:15  1  16.0
  2013-01-01 00:00:18  1  18.5
2 2013-01-01 00:00:18  2  20.0
  2013-01-01 00:00:21  2  22.0
  2013-01-01 00:00:24  2  25.0
  2013-01-01 00:00:27  2  28.0
  2013-01-01 00:00:30  2  30.5
3 2013-01-01 00:00:30  3  32.0
  2013-01-01 00:00:33  3  34.0
  2013-01-01 00:00:36  3  37.0
  2013-01-01 00:00:39  3  39.0

In [4]: df.groupby('A').rolling(3).mean()
Out[4]:
                       A     B
2013-01-01 00:00:00  NaN   NaN
2013-01-01 00:00:01  NaN   NaN
2013-01-01 00:00:02  1.0   1.0
2013-01-01 00:00:03  1.0   2.0
2013-01-01 00:00:04  1.0   3.0
2013-01-01 00:00:05  1.0   4.0
2013-01-01 00:00:06  1.0   5.0
2013-01-01 00:00:07  1.0   6.0
2013-01-01 00:00:08  1.0   7.0
2013-01-01 00:00:09  1.0   8.0
2013-01-01 00:00:10  1.0   9.0
2013-01-01 00:00:11  1.0  10.0
2013-01-01 00:00:12  1.0  11.0
2013-01-01 00:00:13  1.0  12.0
2013-01-01 00:00:14  1.0  13.0
2013-01-01 00:00:15  1.0  14.0
2013-01-01 00:00:16  1.0  15.0
2013-01-01 00:00:17  1.0  16.0
2013-01-01 00:00:18  1.0  17.0
2013-01-01 00:00:19  1.0  18.0
2013-01-01 00:00:20  NaN   NaN
2013-01-01 00:00:21  NaN   NaN
2013-01-01 00:00:22  2.0  21.0
2013-01-01 00:00:23  2.0  22.0
2013-01-01 00:00:24  2.0  23.0
2013-01-01 00:00:25  2.0  24.0
2013-01-01 00:00:26  2.0  25.0
2013-01-01 00:00:27  2.0  26.0
2013-01-01 00:00:28  2.0  27.0
2013-01-01 00:00:29  2.0  28.0
2013-01-01 00:00:30  2.0  29.0
2013-01-01 00:00:31  2.0  30.0
2013-01-01 00:00:32  NaN   NaN
2013-01-01 00:00:33  NaN   NaN
2013-01-01 00:00:34  3.0  33.0
2013-01-01 00:00:35  3.0  34.0
2013-01-01 00:00:36  3.0  35.0
2013-01-01 00:00:37  3.0  36.0
2013-01-01 00:00:38  3.0  37.0
2013-01-01 00:00:39  3.0  38.0

In [5]: pd.__version__
Out[5]: '0.18.0+109.g54e22be'
```

**Comment by jreback** [comment]
@jorisvandenbossche you updated? (I am slightly ahead as merged a couple of things), but this was fixed a while ago

```
In [1]: df = pd.DataFrame({'A': [1] * 20 + [2] * 12 + [3] * 8,'B': np.arange(40)})

In [2]: In [2]: df.index = pd.date_range('20130101',freq='s',periods=40)

In [3]: In [3]: df.groupby('A').resample('3s').mean()
Out[3]: 
                         A     B
A                               
1 2013-01-01 00:00:00  1.0   1.0
  2013-01-01 00:00:03  1.0   4.0
  2013-01-01 00:00:06  1.0   7.0
  2013-01-01 00:00:09  1.0  10.0
  2013-01-01 00:00:12  1.0  13.0
  2013-01-01 00:00:15  1.0  16.0
  2013-01-01 00:00:18  1.0  18.5
2 2013-01-01 00:00:18  2.0  20.0
  2013-01-01 00:00:21  2.0  22.0
  2013-01-01 00:00:24  2.0  25.0
  2013-01-01 00:00:27  2.0  28.0
  2013-01-01 00:00:30  2.0  30.5
3 2013-01-01 00:00:30  3.0  32.0
  2013-01-01 00:00:33  3.0  34.0
  2013-01-01 00:00:36  3.0  37.0
  2013-01-01 00:00:39  3.0  39.0
In [4]: df.groupby('A').rolling(3).mean()
Out[4]: 
                         A     B
A                               
1 2013-01-01 00:00:00  NaN   NaN
  2013-01-01 00:00:01  NaN   NaN
  2013-01-01 00:00:02  1.0   1.0
  2013-01-01 00:00:03  1.0   2.0
  2013-01-01 00:00:04  1.0   3.0
  2013-01-01 00:00:05  1.0   4.0
  2013-01-01 00:00:06  1.0   5.0
  2013-01-01 00:00:07  1.0   6.0
  2013-01-01 00:00:08  1.0   7.0
  2013-01-01 00:00:09  1.0   8.0
  2013-01-01 00:00:10  1.0   9.0
  2013-01-01 00:00:11  1.0  10.0
  2013-01-01 00:00:12  1.0  11.0
  2013-01-01 00:00:13  1.0  12.0
  2013-01-01 00:00:14  1.0  13.0
  2013-01-01 00:00:15  1.0  14.0
  2013-01-01 00:00:16  1.0  15.0
  2013-01-01 00:00:17  1.0  16.0
  2013-01-01 00:00:18  1.0  17.0
  2013-01-01 00:00:19  1.0  18.0
2 2013-01-01 00:00:20  NaN   NaN
  2013-01-01 00:00:21  NaN   NaN
  2013-01-01 00:00:22  2.0  21.0
  2013-01-01 00:00:23  2.0  22.0
  2013-01-01 00:00:24  2.0  23.0
  2013-01-01 00:00:25  2.0  24.0
  2013-01-01 00:00:26  2.0  25.0
  2013-01-01 00:00:27  2.0  26.0
  2013-01-01 00:00:28  2.0  27.0
  2013-01-01 00:00:29  2.0  28.0
  2013-01-01 00:00:30  2.0  29.0
  2013-01-01 00:00:31  2.0  30.0
3 2013-01-01 00:00:32  NaN   NaN
  2013-01-01 00:00:33  NaN   NaN
  2013-01-01 00:00:34  3.0  33.0
  2013-01-01 00:00:35  3.0  34.0
  2013-01-01 00:00:36  3.0  35.0
  2013-01-01 00:00:37  3.0  36.0
  2013-01-01 00:00:38  3.0  37.0
  2013-01-01 00:00:39  3.0  38.0

In [5]: pd.__version__
Out[5]: '0.18.0+120.g094c425'
```

**Comment by jreback** [comment]
pushed again just in case

**Comment by TomAugspurger** [comment]
@jreback one more lost index name, when you group a Series by another (using the original)

```python
In [27]: df['B'].groupby(df.A).rolling(4).mean()
Out[27]:
A  idx
1  0       NaN
   1       NaN
   2       NaN
   3       1.5
   4       2.5
          ...
3  35     33.5
   36     34.5
   37     35.5
   38     36.5
   39     37.5
dtype: float64
```

Let me know if you want to fix that here, otherwise I'll open an issue.

**Comment by TomAugspurger** [comment]
And I'm (correctly) seeing a MultiIndex for the `df.groupby('A').resample('3s').mean()`

**Comment by jorisvandenbossche** [comment]
Will try again (I just fetched the head of this PR, didn't rebase myself as I thought you already did that)

**Comment by jorisvandenbossche** [comment]
@TomAugspurger It's with the rolling case I do not see a MultiIndex (`df.groupby('A').rolling(3).mean()`), `resample` did work for me as well.

@jreback Tested it again (fetched this PR, rebased on latest master myself, rebuild pandas to be certain), and still getting the same result as I showed in https://github.com/pydata/pandas/pull/12743#issuecomment-210904108 (using python 2.7, numpy 1.10.1, Windows)

**Comment by jreback** [comment]
@jorisvandenbossche hmm, ok I did on 2.7/1.10.4 and it doesn't look right. odd.

**Comment by jreback** [comment]
ok, this has been failing on windows on my branch : https://ci.appveyor.com/project/jreback/pandas/build/1.0.2077/job/qfrsv7p46f6ns6xg

**Comment by jreback** [comment]
ok, this is not a 2.7/numpy issue, but something specific to windows.

**Comment by jreback** [comment]
ok pushed a commit to fix. This comes back to some of the 'figuring what the shape of the return' is logic. IOW its a heuristic depending on whether something is mutated (this is the standard case). But when we are doing a chained window/resample we need to force this.

**Comment by jreback** [comment]
pushed a new version, had some flake/formatting issues on windows.

**Comment by jreback** [comment]
@TomAugspurger I picked up that last name fix (was a bug in the concat step).

**Comment by jreback** [comment]
@TomAugspurger @jorisvandenbossche any more comments.

**Comment by TomAugspurger** [comment]
Nothing else from me 👍

**Comment by jreback** [comment]
ok will merge shortly unless @jorisvandenbossche has any further comments.

**Comment by jorisvandenbossche** [comment]
Tested on master, and now indeed the inconsistency is solved! Thanks

**Comment by bigquant** [comment]
@jreback, thanks.

I want to do grouby then rolling then corr on two columns, code below takes 18s+ on 1M rows:

```
df.groupby('name', as_index=False, sort=False, group_keys=False).apply(
        lambda x: x['_a'].rolling(d).corr(other=x['_b'], pairwise=True))
```

**Is there a faster way to do this**? i tried code but it does not work (got exceptions):

```
_g = df.groupby('name', as_index=False, sort=False, group_keys=False)
_g['open'].rolling(5).corr(other=_g['close'], pairwise=False)              
```

**Pull request closed without merging**